src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+            // std::next(it) may be deleted in the process, so this needs to be decided beforehand.
+            auto it_next = (std::next(it) == index.end() || std::next(it)->m_peer != peer) ?
+                index.end() : std::next(it);
+            // If the entry isn't already COMPLETED, first make it COMPLETED (which will mark other CANDIDATEs as
+            // CANDIDATE_BEST, or delete all of a txhash's entries if no non-COMPLETED ones are left).
+            if (MakeCompleted(m_index.project<ByTxHash>(it))) {
+                // Then actually delete the entry (unless it was already deleted by MakeCompleted).
+                Erase<ByPeer>(it);
+            }
+            it = it_next;
+        }
+    }
+
+    void ForgetTxHash(const uint256& txhash)
+    {
+        auto it = m_index.get<ByTxHash>().lower_bound(EntryTxHash{txhash, State::CANDIDATE_DELAYED, 0});
+        while (it != m_index.get<ByTxHash>().end() && it->m_txhash == txhash) {
+            it = Erase<ByTxHash>(it);
+        }
+    }
+
+    void ReceivedInv(uint64_t peer, const GenTxid& gtxid, bool preferred,
+        std::chrono::microseconds reqtime)
+    {
+        // Bail out if we already have a CANDIDATE_BEST entry for this (txhash, peer) combination. The case where
+        // there is a non-CANDIDATE_BEST entry already will be caught by the uniqueness property of the ByPeer index
+        // automatically.
+        if (m_index.get<ByPeer>().count(EntryPeer{peer, true, gtxid.GetHash()})) return;
+
+        // Find last entry for this txhash, and extract per-txhash information from it.
+        Iter<ByTxHash> it_last = m_index.get<ByTxHash>().end();
+        // First find the first entry past this txhash.
+        it_last = m_index.get<ByTxHash>().lower_bound(EntryTxHash{gtxid.GetHash(), State::TOO_LARGE, 0});
+        if (it_last != m_index.get<ByTxHash>().begin() && std::prev(it_last)->m_txhash == gtxid.GetHash()) {
+            it_last--;
+        } else {
+            // No entry for this txhash exists yet.
+            it_last = m_index.get<ByTxHash>().end();
+        }
+
+        // Try creating the entry with CANDIDATE_DELAYED state (which will fail due to the uniqueness
+        // of the ByPeer index if a non-CANDIDATE_BEST entry already exists with the same txhash and peer).
+        // Bail out in that case.
+        auto ret = m_index.get<ByPeer>().emplace(gtxid, peer, preferred, reqtime, m_sequence);
+        if (!ret.second) return;
+
+        // Update accounting metadata.
+        ++m_peerinfo[peer].m_total;
+        ++m_sequence;
+    }
+
+    //! Find the GenTxids to request now from peer.
+    std::vector<GenTxid> GetRequestable(uint64_t peer, std::chrono::microseconds now)
+    {
+        // Move time.
+        SetTimePoint(now);
+
+        // Find all CANDIDATE_BEST entries for this peer.
+        std::vector<std::pair<uint64_t, const Entry*>> selected;
```

Comment:
Is there a reason not to just construct the `GenTxid`s directly into `selected`?

Reply:
It needs to be sorted by sequence before being returned to ensure txs are requested in order of announcement, and the sequence isn't needed in the return value, so having a temporary vector's sensible, I think. And sorting a vector of 16-byte seq/pointer pairs should be more cache efficient than for 48-byte seq/is_wtxid/uint256 tuples (or 40 bytes if you combined is_wtxid into seq).

Reply:
Yeah, the reason is that we want to sort by sequence number, but those aren't included in the output itself, so we need some kind of proxy. It could be a list of (sequence, gtxid) pairs that are sorted, but that would still require an extraction step to convert it to just gtxids. This seems simplest.

Reply:
Yeah, I understood the sorting step, and was just wondering why not create a vector of (seq, gtxid) pairs. As you point out, you'd still need an extraction (copy) step afterwards, so that seems less efficient.